File: facebook-core/src/main/java/com/facebook/appevents/codeless/CodelessMatcher.java
/*
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * You are hereby granted a non-exclusive, worldwide, royalty-free license to use,
 * copy, modify, and distribute this software in source code or binary form for use
 * in connection with the web services and APIs provided by Facebook.
 *
 * As with any software that integrates with the Facebook platform, your use of
 * this software is subject to the Facebook Developer Principles and Policies
 * [http://developers.facebook.com/policy/]. This copyright notice shall be
 * included in all copies or substantial portions of the software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package com.facebook.appevents.codeless;

import android.app.Activity;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.support.annotation.Nullable;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewParent;
import android.view.ViewTreeObserver;

import com.facebook.FacebookException;
import com.facebook.FacebookSdk;
import com.facebook.appevents.codeless.internal.Constants;
import com.facebook.appevents.codeless.internal.ViewHierarchy;
import com.facebook.appevents.codeless.internal.ParameterComponent;
import com.facebook.appevents.codeless.internal.PathComponent;
import com.facebook.appevents.codeless.internal.EventBinding;
import com.facebook.internal.FetchedAppSettings;
import com.facebook.internal.FetchedAppSettingsManager;
import com.facebook.internal.InternalSettings;
import com.facebook.internal.Utility;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import static com.facebook.appevents.codeless.internal.PathComponent.MatchBitmaskType;

class CodelessMatcher {
    private static final String PARENT_CLASS_NAME = "..";
    private static final String CURRENT_CLASS_NAME = ".";
    private static final String TAG = CodelessMatcher.class.getCanonicalName();

    private final Handler uiThreadHandler;
    private Set<Activity> activitiesSet;
    private Set<ViewMatcher> viewMatchers;
    private HashSet<String> listenerSet;
    private HashMap<Integer, HashSet<String>> activityToListenerMap;

    private static CodelessMatcher codelessMatcher = null;

    private CodelessMatcher() {
        this.uiThreadHandler = new Handler(Looper.getMainLooper());
        this.activitiesSet = new HashSet<>();
        this.viewMatchers = new HashSet<>();
        this.listenerSet = new HashSet<>();
        this.activityToListenerMap = new HashMap<>();
    }

    public static synchronized CodelessMatcher getInstance() {
        if (codelessMatcher == null) {
            codelessMatcher = new CodelessMatcher();
        }
        return codelessMatcher;
    }

    public void add(Activity activity) {
        if (InternalSettings.isUnityApp()) {
            return;
        }
        if (Thread.currentThread() != Looper.getMainLooper().getThread()) {
            throw new FacebookException("Can't add activity to CodelessMatcher on non-UI thread");
        }
        this.activitiesSet.add(activity);
        listenerSet.clear();
        if (activityToListenerMap.containsKey(activity.hashCode())) {
            listenerSet = activityToListenerMap.get(activity.hashCode());
        }
        startTracking();
    }

    public void remove(Activity activity) {
        if (InternalSettings.isUnityApp()) {
            return;
        }
        if (Thread.currentThread() != Looper.getMainLooper().getThread()) {
            throw new FacebookException(
                    "Can't remove activity from CodelessMatcher on non-UI thread"
            );
        }
        this.activitiesSet.remove(activity);
        this.viewMatchers.clear();
        this.activityToListenerMap.put(activity.hashCode(),
            (HashSet<String>) listenerSet.clone());
        listenerSet.clear();
    }

    public void destroy(Activity activity) {
        this.activityToListenerMap.remove(activity.hashCode());
    }

    public static Bundle getParameters(final EventBinding mapping,
                                       final View rootView,
                                       final View hostView) {
        Bundle params = new Bundle();

        if (null == mapping) {
            return params;
        }

        List<ParameterComponent> parameters = mapping.getViewParameters();
        if (null != parameters) {
            for (ParameterComponent component : parameters) {
                if (component.value != null && component.value.length() > 0) {
                    params.putString(component.name, component.value);
                } else if (component.path.size() > 0){
                    List<MatchedView> matchedViews;
                    final String pathType = component.pathType;
                    if (pathType.equals(Constants.PATH_TYPE_RELATIVE)) {
                        matchedViews = ViewMatcher.findViewByPath(
                                mapping,
                                hostView,
                                component.path,
                                0,
                                -1,
                                hostView.getClass().getSimpleName()
                        );
                    } else {
                        matchedViews = ViewMatcher.findViewByPath(
                                mapping,
                                rootView,
                                component.path,
                                0,
                                -1,
                                rootView.getClass().getSimpleName()
                        );
                    }

                    for (MatchedView view : matchedViews) {
                        if (view.getView() == null) {
                            continue;
                        }
                        String text = ViewHierarchy.getTextOfView(view.getView());
                        if (text.length() > 0) {
                            params.putString(component.name, text);
                            break;
                        }
                    }
                }
            }
        }

        return params;
    }

    private void startTracking() {
        if (Thread.currentThread() == Looper.getMainLooper().getThread()) {
            matchViews();
        } else {
            uiThreadHandler.post(new Runnable() {
                @Override
                public void run() {
                    matchViews();
                }
            });

        }
    }

    private void matchViews() {
        for (Activity activity : this.activitiesSet) {
            final View rootView = activity.getWindow().getDecorView().getRootView();
            final String activityName = activity.getClass().getSimpleName();
            ViewMatcher matcher = new ViewMatcher(
                    rootView, uiThreadHandler, listenerSet, activityName);
            this.viewMatchers.add(matcher);
        }
    }

    public static class MatchedView {
        private WeakReference<View> view;
        private String viewMapKey;

        public MatchedView(View view, String viewMapKey) {
            this.view = new WeakReference<View>(view);
            this.viewMapKey = viewMapKey;
        }

        @Nullable
        public View getView() {
            return (this.view == null) ? null : this.view.get();
        }

        public String getViewMapKey() {
            return this.viewMapKey;
        }
    }

    protected static class ViewMatcher implements ViewTreeObserver.OnGlobalLayoutListener,
            ViewTreeObserver.OnScrollChangedListener, Runnable {
        private WeakReference<View> rootView;
        @Nullable private List<EventBinding> eventBindings;
        private final Handler handler;
        private HashSet<String> listenerSet;
        private final String activityName;

        public ViewMatcher(View rootView,
                           Handler handler,
                           HashSet<String> listenerSet,
                           final String activityName) {
            this.rootView = new WeakReference<View>(rootView);
            this.handler = handler;
            this.listenerSet = listenerSet;
            this.activityName = activityName;

            this.handler.postDelayed(this, 200);
        }

        @Override
        public void run() {
            final String appId = FacebookSdk.getApplicationId();
            FetchedAppSettings appSettings =
                    FetchedAppSettingsManager.getAppSettingsWithoutQuery(appId);
            if (appSettings == null || !appSettings.getCodelessEventsEnabled()) {
                return;
            }

            this.eventBindings = EventBinding.parseArray(appSettings.getEventBindings());

            if (this.eventBindings != null) {
                View rootView = this.rootView.get();
                if (rootView == null) {
                    return;
                }
                ViewTreeObserver observer = rootView.getViewTreeObserver();
                if (observer.isAlive()) {
                    observer.addOnGlobalLayoutListener(this);
                    observer.addOnScrollChangedListener(this);
                }

                startMatch();
            }
        }

        @Override
        public void onGlobalLayout() {
            startMatch();
        }

        @Override
        public void onScrollChanged() {
            startMatch();
        }

        private void startMatch() {
            if (this.eventBindings != null && this.rootView.get() != null) {
                for (int i = 0; i < this.eventBindings.size(); i++) {
                    EventBinding binding = this.eventBindings.get(i);
                    findView(binding, this.rootView.get());
                }
            }
        }

        public void findView(final EventBinding mapping, final View rootView) {
            if (mapping == null || rootView == null) {
                return;
            }

            if (!TextUtils.isEmpty(mapping.getActivityName()) &&
                    !mapping.getActivityName().equals(this.activityName)) {
                return;
            }

            List<PathComponent> path = mapping.getViewPath();

            if (path.size() > Constants.MAX_TREE_DEPTH) {
                return;
            }

            List<MatchedView> matchedViews = findViewByPath(
                    mapping,
                    rootView,
                    path,
                    0,
                    -1,
                    this.activityName);
            for (MatchedView view: matchedViews) {
                attachListener(view, rootView, mapping);
            }
        }

        public static List<MatchedView> findViewByPath(final EventBinding mapping,
                                   final View view,
                                   final List<PathComponent> path,
                                   final int level,
                                   final int index,
                                   String mapKey) {
            mapKey += "." + String.valueOf(index);
            List<MatchedView> result = new ArrayList<>();
            if (null == view) {
                return result;
            }

            if (level >= path.size()) {
                // Match all children views if their parent view is matched
                result.add(new MatchedView(view, mapKey));
            } else {
                PathComponent pathElement = path.get(level);
                if (pathElement.className.equals(PARENT_CLASS_NAME)) {
                    ViewParent parent = view.getParent();
                    if (parent instanceof ViewGroup) {
                        final ViewGroup viewGroup = (ViewGroup)parent;
                        List<View> visibleViews = findVisibleChildren(viewGroup);
                        final int childCount = visibleViews.size();
                        for (int i = 0; i < childCount; i++) {
                            View child = visibleViews.get(i);
                            List<MatchedView> matchedViews = findViewByPath(
                                    mapping,
                                    child,
                                    path,
                                    level + 1,
                                    i,
                                    mapKey);
                            result.addAll(matchedViews);
                        }
                    }

                    return result;
                } else if (pathElement.className.equals(CURRENT_CLASS_NAME)) {
                    // Set self as selected element
                    result.add(new MatchedView(view, mapKey));

                    return result;
                }

                if (!isTheSameView(view, pathElement, index)) {
                    return result;
                }

                // Found it!
                if (level == path.size() - 1) {
                    result.add(new MatchedView(view, mapKey));
                }
            }

            if (view instanceof ViewGroup) {
                final ViewGroup viewGroup = (ViewGroup) view;
                List<View> visibleViews = findVisibleChildren(viewGroup);
                final int childCount = visibleViews.size();
                for (int i = 0; i < childCount; i++) {
                    View child = visibleViews.get(i);
                    List<MatchedView> matchedViews = findViewByPath(
                            mapping,
                            child,
                            path,
                            level + 1,
                            i,
                            mapKey);
                    result.addAll(matchedViews);
                }
            }

            return result;
        }

        private static boolean isTheSameView(
                final View targetView,
                final PathComponent pathElement,
                final int index) {
            if (pathElement.index != -1 && index != pathElement.index) {
                return false;
            }

            if (!targetView.getClass().getCanonicalName().equals(pathElement.className)) {
                if (pathElement.className.matches(".*android\\..*")) {
                    String[] names = pathElement.className.split("\\.");
                    if (names.length > 0) {
                        String SimpleName = names[names.length - 1];
                        if (!targetView.getClass().getSimpleName().equals(SimpleName)) {
                            return false;
                        }
                    } else {
                        return false;
                    }
                } else {
                    return false;
                }
            }

            if ((pathElement.matchBitmask
                    & MatchBitmaskType.ID.getValue()) > 0) {
                if (pathElement.id != targetView.getId()) {
                    return false;
                }
            }

            if ((pathElement.matchBitmask
                    & MatchBitmaskType.TEXT.getValue()) > 0) {
                String pathText = pathElement.text;
                String text = ViewHierarchy.getTextOfView(targetView);
                String hashedText = Utility.coerceValueIfNullOrEmpty(
                        Utility.sha256hash(text), "");

                if (!pathText.equals(text) && !pathText.equals(hashedText)) {
                    return false;
                }
            }

            if ((pathElement.matchBitmask
                    & MatchBitmaskType.DESCRIPTION.getValue()) > 0) {
                String pathDesc = pathElement.description;
                String targetDesc = targetView.getContentDescription() == null ? "" :
                        String.valueOf(targetView.getContentDescription());
                String hashedDesc = Utility.coerceValueIfNullOrEmpty(
                        Utility.sha256hash(targetDesc), "");
                if (!pathDesc.equals(targetDesc) && !pathDesc.equals(hashedDesc)) {
                    return false;
                }
            }

            if ((pathElement.matchBitmask
                    & MatchBitmaskType.HINT.getValue()) > 0) {
                String pathHint = pathElement.hint;
                String targetHint = ViewHierarchy.getHintOfView(targetView);
                String hashedHint = Utility.coerceValueIfNullOrEmpty(
                        Utility.sha256hash(targetHint), "");

                if (!pathHint.equals(targetHint) && !pathHint.equals(hashedHint)) {
                    return false;
                }
            }

            if ((pathElement.matchBitmask
                    & MatchBitmaskType.TAG.getValue()) > 0) {
                String tag = pathElement.tag;
                String targetTag = targetView.getTag() == null ? "" :
                        String.valueOf(targetView.getTag());
                String hashedTag = Utility.coerceValueIfNullOrEmpty(
                        Utility.sha256hash(targetTag), "");
                if (!tag.equals(targetTag) && !tag.equals(hashedTag)) {
                    return false;
                }
            }

            return true;
        }

        private static List<View> findVisibleChildren(ViewGroup viewGroup) {
            List<View> visibleViews = new ArrayList<>();
            final int childCount = viewGroup.getChildCount();
            for (int i = 0; i < childCount; i++) {
                View child = viewGroup.getChildAt(i);
                if (child.getVisibility() == View.VISIBLE) {
                    visibleViews.add(child);
                }
            }
            return visibleViews;
        }

        private void attachListener(final MatchedView matchedView,
                                    final View rootView,
                                    final EventBinding mapping) {
            if (mapping == null) {
                return;
            }
            try {
                View view = matchedView.getView();
                if (view == null) {
                    return;
                }

                // If it's React Native Button, then attach React Native OnTouchListener
                View RCTRootView = ViewHierarchy.findRCTRootView(view);
                if (null != RCTRootView && ViewHierarchy.isRCTButton(view, RCTRootView)) {
                    attachRCTListener(matchedView, rootView, RCTRootView, mapping);
                    return;
                }

                // Skip if the view comes from React Native
                if (view.getClass().getName().startsWith("com.facebook.react")) {
                    return;
                }

                final String mapKey = matchedView.getViewMapKey();
                View.OnClickListener existingListener =
                        ViewHierarchy.getExistingOnClickListener(view);
                boolean listenerExists = existingListener != null;
                boolean isCodelessListener = listenerExists && existingListener instanceof
                        CodelessLoggingEventListener.AutoLoggingOnClickListener;
                boolean listenerSupportCodelessLogging = isCodelessListener &&
                        ((CodelessLoggingEventListener.AutoLoggingOnClickListener)
                                existingListener).getSupportCodelessLogging();
                if (!this.listenerSet.contains(mapKey) &&
                        (!listenerExists ||
                                !isCodelessListener || !listenerSupportCodelessLogging)) {
                    View.OnClickListener listener =
                            CodelessLoggingEventListener.getOnClickListener(
                                    mapping, rootView, view);
                    view.setOnClickListener(listener);
                    this.listenerSet.add(mapKey);
                }
            } catch (FacebookException e) {
                Log.e(TAG, "Failed to attach auto logging event listener.", e);
            }
        }

        private void attachRCTListener(final MatchedView matchedView,
                                       final View rootView,
                                       final View RCTRootView,
                                       final EventBinding mapping){
            if (mapping == null) {
                return;
            }
            View view = matchedView.getView();
            if (view == null || !ViewHierarchy.isRCTButton(view, RCTRootView)) {
                return;
            }

            final String mapKey = matchedView.getViewMapKey();
            View.OnTouchListener existingListener =
                    ViewHierarchy.getExistingOnTouchListener(view);
            boolean listenerExists = existingListener != null;
            boolean isCodelessListener = listenerExists && existingListener instanceof
                    RCTCodelessLoggingEventListener.AutoLoggingOnTouchListener;
            boolean listenerSupportCodelessLogging = isCodelessListener &&
                    ((RCTCodelessLoggingEventListener.AutoLoggingOnTouchListener)
                            existingListener).getSupportCodelessLogging();
            if (!this.listenerSet.contains(mapKey) &&
                    (!listenerExists ||
                            !isCodelessListener || !listenerSupportCodelessLogging)) {
                View.OnTouchListener listener =
                        RCTCodelessLoggingEventListener.getOnTouchListener(
                                mapping, rootView, view);
                view.setOnTouchListener(listener);
                this.listenerSet.add(mapKey);
            }
        }
    }
}

File: facebook-core/src/main/java/com/facebook/appevents/codeless/CodelessMatcher.kt
/*
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * You are hereby granted a non-exclusive, worldwide, royalty-free license to use,
 * copy, modify, and distribute this software in source code or binary form for use
 * in connection with the web services and APIs provided by Facebook.
 *
 * As with any software that integrates with the Facebook platform, your use of
 * this software is subject to the Facebook Developer Principles and Policies
 * [http://developers.facebook.com/policy/]. This copyright notice shall be
 * included in all copies or substantial portions of the software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package com.facebook.appevents.codeless

import android.app.Activity
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.View
import android.view.ViewGroup
import android.view.ViewTreeObserver
import android.widget.AdapterView
import android.widget.ListView
import androidx.annotation.UiThread
import com.facebook.FacebookException
import com.facebook.FacebookSdk.getApplicationId
import com.facebook.appevents.codeless.CodelessLoggingEventListener.AutoLoggingOnClickListener
import com.facebook.appevents.codeless.CodelessLoggingEventListener.AutoLoggingOnItemClickListener
import com.facebook.appevents.codeless.internal.Constants
import com.facebook.appevents.codeless.internal.EventBinding
import com.facebook.appevents.codeless.internal.PathComponent
import com.facebook.appevents.codeless.internal.ViewHierarchy
import com.facebook.appevents.internal.AppEventUtility.getRootView
import com.facebook.internal.FetchedAppSettingsManager.getAppSettingsWithoutQuery
import com.facebook.internal.InternalSettings.isUnityApp
import com.facebook.internal.Utility.coerceValueIfNullOrEmpty
import com.facebook.internal.Utility.logd
import com.facebook.internal.Utility.sha256hash
import com.facebook.internal.instrument.crashshield.AutoHandleExceptions
import java.lang.ref.WeakReference
import java.util.Collections
import java.util.WeakHashMap
import kotlin.collections.ArrayList
import kotlin.collections.HashMap
import kotlin.collections.HashSet

@AutoHandleExceptions
internal class CodelessMatcher private constructor() {
  private val uiThreadHandler: Handler = Handler(Looper.getMainLooper())
  private val activitiesSet: MutableSet<Activity> = Collections.newSetFromMap(WeakHashMap())
  private val viewMatchers: MutableSet<ViewMatcher> = mutableSetOf()
  private var listenerSet: HashSet<String> = hashSetOf()
  private val activityToListenerMap: HashMap<Int, HashSet<String>> = hashMapOf()
  @UiThread
  fun add(activity: Activity) {
    if (isUnityApp) {
      return
    }
    if (Thread.currentThread() !== Looper.getMainLooper().thread) {
      throw FacebookException("Can't add activity to CodelessMatcher on non-UI thread")
    }
    activitiesSet.add(activity)
    listenerSet.clear()
    activityToListenerMap[activity.hashCode()]?.let { listenerSet = it }
    startTracking()
  }

  @UiThread
  fun remove(activity: Activity) {
    if (isUnityApp) {
      return
    }
    if (Thread.currentThread() !== Looper.getMainLooper().thread) {
      throw FacebookException("Can't remove activity from CodelessMatcher on non-UI thread")
    }
    activitiesSet.remove(activity)
    viewMatchers.clear()
    activityToListenerMap[activity.hashCode()] = listenerSet.clone() as HashSet<String>
    listenerSet.clear()
  }

  @UiThread
  fun destroy(activity: Activity) {
    activityToListenerMap.remove(activity.hashCode())
  }

  private fun startTracking() {
    if (Thread.currentThread() === Looper.getMainLooper().thread) {
      matchViews()
    } else {
      uiThreadHandler.post { matchViews() }
    }
  }

  private fun matchViews() {
    for (activity in activitiesSet) {
      if (null != activity) {
        val rootView = getRootView(activity)
        val activityName = activity.javaClass.simpleName
        val matcher = ViewMatcher(rootView, uiThreadHandler, listenerSet, activityName)
        viewMatchers.add(matcher)
      }
    }
  }

  class MatchedView(view: View, viewMapKey: String) {
    private val view: WeakReference<View>?
    val viewMapKey: String
    fun getView(): View? {
      return view?.get()
    }

    init {
      this.view = WeakReference(view)
      this.viewMapKey = viewMapKey
    }
  }

  @UiThread
  class ViewMatcher(
      rootView: View?,
      handler: Handler,
      listenerSet: HashSet<String>,
      activityName: String
  ) : ViewTreeObserver.OnGlobalLayoutListener, ViewTreeObserver.OnScrollChangedListener, Runnable {
    private val rootView: WeakReference<View?> = WeakReference(rootView)
    private var eventBindings: List<EventBinding>? = null
    private val handler: Handler = handler
    private val listenerSet: HashSet<String> = listenerSet
    private val activityName: String = activityName
    override fun run() {
      val appId = getApplicationId()
      val appSettings = getAppSettingsWithoutQuery(appId)
      if (appSettings == null || !appSettings.codelessEventsEnabled) {
        return
      }
      eventBindings = EventBinding.parseArray(appSettings.eventBindings)
      if (eventBindings != null) {
        val rootView = rootView.get() ?: return
        val observer = rootView.viewTreeObserver
        if (observer.isAlive) {
          observer.addOnGlobalLayoutListener(this)
          observer.addOnScrollChangedListener(this)
        }
        startMatch()
      }
    }

    override fun onGlobalLayout() {
      startMatch()
    }

    override fun onScrollChanged() {
      startMatch()
    }

    private fun startMatch() {
      eventBindings?.let {
        if (rootView.get() != null) {
          for (i in it.indices) {
            val binding = it[i]
            findView(binding, rootView.get())
          }
        }
      }
    }

    private fun findView(mapping: EventBinding?, rootView: View?) {
      if (mapping == null || rootView == null) {
        return
      }
      if (!mapping.activityName.isNullOrEmpty() && mapping.activityName != activityName) {
        return
      }
      val path = mapping.viewPath
      if (path.size > Constants.MAX_TREE_DEPTH) {
        return
      }
      val matchedViews = findViewByPath(mapping, rootView, path, 0, -1, activityName)
      for (view in matchedViews) {
        attachListener(view, rootView, mapping)
      }
    }

    private fun attachListener(matchedView: MatchedView, rootView: View, mapping: EventBinding?) {
      if (mapping == null) {
        return
      }
      try {
        val view = matchedView.getView() ?: return
        // If it's React Native Button, then attach React Native OnTouchListener
        val rctRootView = ViewHierarchy.findRCTRootView(view)
        if (null != rctRootView && ViewHierarchy.isRCTButton(view, rctRootView)) {
          attachRCTListener(matchedView, rootView, mapping)
          return
        }
        // Skip if the view comes from React Native
        if (view.javaClass.name.startsWith("com.facebook.react")) {
          return
        }
        if (view !is AdapterView<*>) {
          // attach onClickListener
          attachOnClickListener(matchedView, rootView, mapping)
        } else if (view is ListView) {
          // attach AdapterView onItemClickListener
          attachOnItemClickListener(matchedView, rootView, mapping)
        }
      } catch (e: Exception) {
        logd(TAG, e)
      }
    }

    private fun attachOnClickListener(
        matchedView: MatchedView,
        rootView: View,
        mapping: EventBinding
    ) {
      val view = matchedView.getView() ?: return
      val mapKey = matchedView.viewMapKey
      val existingListener = ViewHierarchy.getExistingOnClickListener(view)
      val isCodelessListener = existingListener is AutoLoggingOnClickListener
      val listenerSupportCodelessLogging =
          isCodelessListener &&
              (existingListener as AutoLoggingOnClickListener).supportCodelessLogging
      if (!listenerSet.contains(mapKey) && !listenerSupportCodelessLogging) {
        val listener: View.OnClickListener =
            CodelessLoggingEventListener.getOnClickListener(mapping, rootView, view)
        view.setOnClickListener(listener)
        listenerSet.add(mapKey)
      }
    }

    private fun attachOnItemClickListener(
        matchedView: MatchedView,
        rootView: View,
        mapping: EventBinding
    ) {
      val view = matchedView.getView() as AdapterView<*>? ?: return
      val mapKey = matchedView.viewMapKey
      val existingListener = view.onItemClickListener
      val isCodelessListener = existingListener is AutoLoggingOnItemClickListener
      val listenerSupportCodelessLogging =
          isCodelessListener &&
              (existingListener as AutoLoggingOnItemClickListener).supportCodelessLogging
      if (!listenerSet.contains(mapKey) && !listenerSupportCodelessLogging) {
        val listener: AdapterView.OnItemClickListener =
            CodelessLoggingEventListener.getOnItemClickListener(mapping, rootView, view)
        view.onItemClickListener = listener
        listenerSet.add(mapKey)
      }
    }

    private fun attachRCTListener(matchedView: MatchedView, rootView: View, mapping: EventBinding) {
      val view = matchedView.getView() ?: return
      val mapKey = matchedView.viewMapKey
      val existingListener = ViewHierarchy.getExistingOnTouchListener(view)
      val isRCTCodelessListener =
          existingListener is RCTCodelessLoggingEventListener.AutoLoggingOnTouchListener
      val listenerSupportCodelessLogging =
          isRCTCodelessListener &&
              (existingListener as RCTCodelessLoggingEventListener.AutoLoggingOnTouchListener)
                  .supportCodelessLogging
      if (!listenerSet.contains(mapKey) && !listenerSupportCodelessLogging) {
        val listener: View.OnTouchListener =
            RCTCodelessLoggingEventListener.getOnTouchListener(mapping, rootView, view)
        view.setOnTouchListener(listener)
        listenerSet.add(mapKey)
      }
    }

    companion object {
      @JvmStatic
      fun findViewByPath(
          mapping: EventBinding?,
          view: View?,
          path: List<PathComponent>,
          level: Int,
          index: Int,
          mapKey: String
      ): List<MatchedView> {
        val mapKey = "$mapKey.$index"
        val result: MutableList<MatchedView> = ArrayList()
        if (null == view) {
          return result
        }
        if (level >= path.size) {
          // Match all children views if their parent view is matched
          result.add(MatchedView(view, mapKey))
        } else {
          val pathElement = path[level]
          if (pathElement.className == PARENT_CLASS_NAME) {
            val parent = view.parent
            if (parent is ViewGroup) {
              val visibleViews = findVisibleChildren(parent)
              val childCount = visibleViews.size
              for (i in 0 until childCount) {
                val child = visibleViews[i]
                val matchedViews = findViewByPath(mapping, child, path, level + 1, i, mapKey)
                result.addAll(matchedViews)
              }
            }
            return result
          } else if (pathElement.className == CURRENT_CLASS_NAME) {
            // Set self as selected element
            result.add(MatchedView(view, mapKey))
            return result
          }
          if (!isTheSameView(view, pathElement, index)) {
            return result
          }

          // Found it!
          if (level == path.size - 1) {
            result.add(MatchedView(view, mapKey))
          }
        }
        if (view is ViewGroup) {
          val visibleViews = findVisibleChildren(view)
          val childCount = visibleViews.size
          for (i in 0 until childCount) {
            val child = visibleViews[i]
            val matchedViews = findViewByPath(mapping, child, path, level + 1, i, mapKey)
            result.addAll(matchedViews)
          }
        }
        return result
      }

      private fun isTheSameView(targetView: View, pathElement: PathComponent, index: Int): Boolean {
        if (pathElement.index != -1 && index != pathElement.index) {
          return false
        }
        if (targetView.javaClass.canonicalName != pathElement.className) {
          if (pathElement.className.matches(".*android\\..*".toRegex())) {
            val names = pathElement.className.split(".")
            if (names.isNotEmpty()) {
              val simpleName = names[names.size - 1]
              if (targetView.javaClass.simpleName != simpleName) {
                return false
              }
            } else {
              return false
            }
          } else {
            return false
          }
        }
        if (pathElement.matchBitmask and PathComponent.MatchBitmaskType.ID.value > 0) {
          if (pathElement.id != targetView.id) {
            return false
          }
        }
        if (pathElement.matchBitmask and PathComponent.MatchBitmaskType.TEXT.value > 0) {
          val pathText = pathElement.text
          val text = ViewHierarchy.getTextOfView(targetView)
          val hashedText = coerceValueIfNullOrEmpty(sha256hash(text), "")
          if (pathText != text && pathText != hashedText) {
            return false
          }
        }
        if (pathElement.matchBitmask and PathComponent.MatchBitmaskType.DESCRIPTION.value > 0) {
          val pathDesc = pathElement.description
          val targetDesc =
              if (targetView.contentDescription == null) ""
              else targetView.contentDescription.toString()
          val hashedDesc = coerceValueIfNullOrEmpty(sha256hash(targetDesc), "")
          if (pathDesc != targetDesc && pathDesc != hashedDesc) {
            return false
          }
        }
        if (pathElement.matchBitmask and PathComponent.MatchBitmaskType.HINT.value > 0) {
          val pathHint = pathElement.hint
          val targetHint = ViewHierarchy.getHintOfView(targetView)
          val hashedHint = coerceValueIfNullOrEmpty(sha256hash(targetHint), "")
          if (pathHint != targetHint && pathHint != hashedHint) {
            return false
          }
        }
        if (pathElement.matchBitmask and PathComponent.MatchBitmaskType.TAG.value > 0) {
          val tag = pathElement.tag
          val targetTag = if (targetView.tag == null) "" else targetView.tag.toString()
          val hashedTag = coerceValueIfNullOrEmpty(sha256hash(targetTag), "")
          if (tag != targetTag && tag != hashedTag) {
            return false
          }
        }
        return true
      }

      private fun findVisibleChildren(viewGroup: ViewGroup): List<View> {
        val visibleViews: MutableList<View> = ArrayList()
        val childCount = viewGroup.childCount
        for (i in 0 until childCount) {
          val child = viewGroup.getChildAt(i)
          if (child.visibility == View.VISIBLE) {
            visibleViews.add(child)
          }
        }
        return visibleViews
      }
    }

    init {
      this.handler.postDelayed(this, 200)
    }
  }

  companion object {
    private const val PARENT_CLASS_NAME = ".."
    private const val CURRENT_CLASS_NAME = "."
    private val TAG = CodelessMatcher::class.java.canonicalName
    private var codelessMatcher: CodelessMatcher? = null

    @Synchronized
    @JvmStatic
    fun getInstance(): CodelessMatcher {
      if (codelessMatcher == null) {
        codelessMatcher = CodelessMatcher()
      }
      return codelessMatcher as CodelessMatcher
    }

    @UiThread
    @JvmStatic
    fun getParameters(mapping: EventBinding?, rootView: View, hostView: View): Bundle {
      val params = Bundle()
      if (null == mapping) {
        return params
      }
      val parameters = mapping.viewParameters
      if (null != parameters) {
        for (component in parameters) {
          if (component.value != null && component.value.isNotEmpty()) {
            params.putString(component.name, component.value)
          } else if (component.path.size > 0) {
            var matchedViews: List<MatchedView>
            val pathType = component.pathType
            matchedViews =
                if (pathType == Constants.PATH_TYPE_RELATIVE) {
                  ViewMatcher.findViewByPath(
                      mapping, hostView, component.path, 0, -1, hostView.javaClass.simpleName)
                } else {
                  ViewMatcher.findViewByPath(
                      mapping, rootView, component.path, 0, -1, rootView.javaClass.simpleName)
                }
            for (view in matchedViews) {
              if (view.getView() == null) {
                continue
              }
              val text = ViewHierarchy.getTextOfView(view.getView())
              if (text.isNotEmpty()) {
                params.putString(component.name, text)
                break
              }
            }
          }
        }
      }
      return params
    }
  }
}
